sound/soc/sof/intel/hda-dai.c

```diff
@@ -302,6 +303,15 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 	}
 
 	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_STOP:
+		hda_stream = hstream_to_sof_hda_stream(hext_stream);
+
+		/* clean up the DMA for playback if the dma_cleanup_during_stop is set */
+		if (hext_stream->hstream.direction == SNDRV_PCM_STREAM_CAPTURE ||
+		    !hda_stream->dma_cleanup_during_stop)
+			break;
+		hda_stream->dma_cleanup_during_stop = false;
+		fallthrough;
```

Comment:
@ranj063, this is essentially equals to:
```diff
diff --git a/sound/soc/sof/intel/hda-dai-ops.c b/sound/soc/sof/intel/hda-dai-ops.c
index 484c76147885..dffe9e145a1e 100644
--- a/sound/soc/sof/intel/hda-dai-ops.c
+++ b/sound/soc/sof/intel/hda-dai-ops.c
@@ -512,7 +512,6 @@ static const struct hda_dai_widget_dma_ops sdw_ipc4_chain_dma_ops = {
 static int hda_ipc3_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
 				 struct snd_pcm_substream *substream, int cmd)
 {
-	struct hdac_ext_stream *hext_stream = hda_get_hext_stream(sdev, cpu_dai, substream);
 	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
 
 	switch (cmd) {
@@ -527,9 +526,6 @@ static int hda_ipc3_post_trigger(struct snd_sof_dev *sdev, struct snd_soc_dai *c
 		if (ret < 0)
 			return ret;
 
-		if (cmd == SNDRV_PCM_TRIGGER_STOP)
-			return hda_link_dma_cleanup(substream, hext_stream, cpu_dai);
-
 		break;
 	}
 	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
diff --git a/sound/soc/sof/intel/hda-dai.c b/sound/soc/sof/intel/hda-dai.c
index f1a8491bba9e..df8bf497e224 100644
--- a/sound/soc/sof/intel/hda-dai.c
+++ b/sound/soc/sof/intel/hda-dai.c
@@ -303,6 +303,7 @@ static int __maybe_unused hda_dai_trigger(struct snd_pcm_substream *substream, i
 
 	switch (cmd) {
 	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_STOP:
 		ret = hda_link_dma_cleanup(substream, hext_stream, dai);
 		if (ret < 0) {
 			dev_err(sdev->dev, "%s: failed to clean up link DMA\n", __func__);
```

I don't see why we need to exclude capture, apart from that we did not tested. It has the same issue.
At the end we will do the same as we do with IPC3, right? Afaik we have most of these hda-ops complication to _not_ do the same thing for IPC4, there must be a reason why we did this in the first place and diverged from IPC3 flow.

This is one note, the second is: with this change you need to revert (at the same time):
1abc2642588e ("ASoC: SOF: Intel: hda: Compensate LLP in case it is not reset")
f9eeb6bb13fb ("ALSA: hda: Add pplcllpl/u members to hdac_ext_stream")
To not break the delay reporting when the firmware is not reporting LLP (I need to test the firmware reporting case still)

The third and might not be applicable is that we have hda_link_dma_cleanup() called in hda_dai_hw_free(), so we clean up twice, I'm not sure if that is a problem or not, but something to keep in mind.
Right, it might handling the case when PCM is opened, configured, _not_ triggered and then closed?

Reply:
FWIW, the firmware reported LLP works, but I have not tested LNL, where the HD-DMA link LLP is also reported by firmware...

Reply:
> This is one note, the second is: with this change you need to revert (at the same time):
> [1abc264](https://github.com/thesofproject/linux/commit/1abc2642588e06f6180b3fbb21968cf5d0ba9e5f) ("ASoC: SOF: Intel: hda: Compensate LLP in case it is not reset")
> [f9eeb6b](https://github.com/thesofproject/linux/commit/f9eeb6bb13fb5d7af1ea5b74a10b1f8ead962540) ("ALSA: hda: Add pplcllpl/u members to hdac_ext_stream")
> To not break the delay reporting when the firmware is not reporting LLP (I need to test the firmware reporting case still)

@ujfalusi should I really revert the commit. We still dont stop the DMAs during pause. So I should only exclude stop from it right?